### PR TITLE
Add mentee comments feature

### LINKED ISSUE #282: Add mentee comments feature
**Is your feature request related to a problem? Please describe.**
- As a mentor or an admin I should be able to add comments on a mentee

**Describe the solution you'd like**
- Create endpoints to perform the CRUD operations
- Mentee shouldn't be allowed or to access or see the comments

### PR DESCRIPTION
**Pull request by anjula-sack**
## Purpose

- The purpose of this PR is to fix #282 

## Goals

- Mentors and Admins should be able to add/delete comments
## Approach

- Create CRUD enpoints
- Create comment model, service and repository

##  Checklist
- [x] This PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.
- [x] I have read and understood the development best practices guidelines ( http://bit.ly/sef-best-practices )
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: src/main/java/org/sefglobal/scholarx/controller/MenteeController.java [hunk 1]
```diff
@@ -1,30 +1,73 @@
...
     private final MenteeService menteeService;
+    private final CommentService commentService;
 
-    public MenteeController(MenteeService menteeService) {
+    public MenteeController(MenteeService menteeService,CommentService commentService) {
         this.menteeService = menteeService;
+        this.commentService = commentService;
+    }
+
+    @PostMapping("/{id}/comment")
```
Comment (Piumal1999) [inline src/main/java/org/sefglobal/scholarx/controller/MenteeController.java:30]:
```suggestion
    @PostMapping("/{id}/comments")
```

Comment (Piumal1999) [reply src/main/java/org/sefglobal/scholarx/controller/MenteeController.java:30]: Also I think we can directly pass a `Comment` object as the request body

File: src/main/java/org/sefglobal/scholarx/model/Comment.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.Id;
+import javax.persistence.Table;
+import javax.persistence.GenerationType;
+import javax.persistence.ManyToOne;
+
+@Entity
+@Table(name = "comment")
+public class Comment {
```
Comment (Piumal1999) [inline src/main/java/org/sefglobal/scholarx/model/Comment.java:13]: We can extend the `BaseScholarxModel` and remove the id from here

File: src/main/java/org/sefglobal/scholarx/service/CommentService.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+    public List<Comment> getAllMenteeComments(long menteeId, long profileId)
+            throws ResourceNotFoundException, BadRequestException, UnauthorizedException {
+        Optional<Mentee> optionalMentee = menteeRepository.findById(menteeId);
+
+        if (!optionalMentee.isPresent()) {
+            String msg = "Error, Mentee by id: " + menteeId + " doesn't exist.";
+            log.error(msg);
+            throw new ResourceNotFoundException(msg);
+        } else if (optionalMentee.get().getProfile().getId() == profileId) {
```
Comment (Piumal1999) [inline src/main/java/org/sefglobal/scholarx/service/CommentService.java:41]: Throw unauthorized exception if the profile doesn't belong to the assignedMentor or an admin.

File: src/main/java/org/sefglobal/scholarx/service/CommentService.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+    public Comment updateComment(long id, long profileId, String menteeComment)
+            throws ResourceNotFoundException {
+        Optional<Comment> optionalComment = commentRepository.findById(id);
+        if (!optionalComment.isPresent()) {
+            String msg = "Error, Comment with id: " + id + " cannot be deleted. " +
+                    "Comment doesn't exist.";
+            log.error(msg);
+            throw new ResourceNotFoundException(msg);
+        }
```
Comment (Piumal1999) [inline src/main/java/org/sefglobal/scholarx/service/CommentService.java:86]: Check whether the profile id on comment is equal to the profile id taken from authentication

File: src/main/java/org/sefglobal/scholarx/service/CommentService.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+    public void deleteComment(long id)
+            throws ResourceNotFoundException {
+        Optional<Comment> optionalComment = commentRepository.findById(id);
+        if (!optionalComment.isPresent()) {
+            String msg = "Error, Comment with id: " + id + " cannot be deleted. " +
+                    "Comment doesn't exist.";
+            log.error(msg);
+            throw new ResourceNotFoundException(msg);
+        }
```
Comment (Piumal1999) [inline src/main/java/org/sefglobal/scholarx/service/CommentService.java:99]: Here also validate the profile id

**Review by Piumal1999 — COMMENTED** [review]
This is awesome! @anjula-sack
Added some suggestions. We'll have to validate the user's profile to prevent normal users from changing the mentee comments

File: src/main/java/org/sefglobal/scholarx/model/Comment.java (hunk 2 above)
Comment (anjula-sack) [reply src/main/java/org/sefglobal/scholarx/model/Comment.java:13]: oh yeah

File: src/main/java/org/sefglobal/scholarx/service/CommentService.java (hunk 4 above)
Comment (anjula-sack) [reply src/main/java/org/sefglobal/scholarx/service/CommentService.java:86]: ok cool

File: src/main/java/org/sefglobal/scholarx/service/CommentService.java (hunk 5 above)
Comment (anjula-sack) [reply src/main/java/org/sefglobal/scholarx/service/CommentService.java:99]: sure

File: src/main/java/org/sefglobal/scholarx/service/CommentService.java (hunk 3 above)
Comment (anjula-sack) [reply src/main/java/org/sefglobal/scholarx/service/CommentService.java:41]: ok

File: src/main/java/org/sefglobal/scholarx/service/CommentService.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+    public Comment updateComment(long id, long profileId, Comment menteeComment)
+            throws ResourceNotFoundException, UnauthorizedException {
+        Optional<Comment> optionalComment = commentRepository.findById(id);
+        if (!optionalComment.isPresent()) {
+            String msg = "Error, Comment with id: " + id + " cannot be deleted. " +
+                    "Comment doesn't exist.";
+            log.error(msg);
+            throw new ResourceNotFoundException(msg);
+        } else if (menteeComment.getCommented_by().getId() != profileId) {
```
Comment (Piumal1999) [inline src/main/java/org/sefglobal/scholarx/service/CommentService.java:94]:
```suggestion
        } else if (optionalComment.get().getCommented_by().getId() != profileId) {
```

File: src/main/java/org/sefglobal/scholarx/service/CommentService.java (hunk 3 above)
Comment (Piumal1999) [reply src/main/java/org/sefglobal/scholarx/service/CommentService.java:47]: We'll have to use the validation used in line 68-77

File: src/main/java/org/sefglobal/scholarx/model/Comment.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.Table;
+import javax.persistence.ManyToOne;
+
+@Entity
+@Table(name = "comment")
+public class Comment extends BaseScholarxModel {
+    @ManyToOne
+    private Mentee mentee;
```
Comment (Piumal1999) [inline src/main/java/org/sefglobal/scholarx/model/Comment.java:12]: I think we'll have to add the relationship to the Mentee model and Ignore the mentee comments with JsonIgnoreProperties. Does this work properly without adding that?

Comment (anjula-sack) [reply src/main/java/org/sefglobal/scholarx/model/Comment.java:12]: yeah

File: src/main/java/org/sefglobal/scholarx/service/CommentService.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        this.profileRepository = profileRepository;
+        this.mentorRepository = mentorRepository;
+        this.menteeRepository = menteeRepository;
+    }
+
+    public List<Comment> getAllMenteeComments(long menteeId, long profileId)
+            throws ResourceNotFoundException, BadRequestException, UnauthorizedException {
+        Optional<Mentee> optionalMentee = menteeRepository.findById(menteeId);
+        Optional<Profile> optionalProfile = profileRepository.findById(profileId);
+        List<Mentor> optionalMentor = mentorRepository.findAllByProfileId(profileId);
```
Comment (Piumal1999) [inline src/main/java/org/sefglobal/scholarx/service/CommentService.java:43]: Here we don't have to get a mentor list. In line 50 we can directly use,
`optionalMentee.get().getAssignedMentor().getProfile.getId()==profileId`

Same in `addMenteeComment` method

File: src/main/java/org/sefglobal/scholarx/service/CommentService.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        comment.setComment(menteeComment.getComment());
+        comment.setMentee(optionalMentee.get());
+        return commentRepository.save(comment);
+    }
+
+    public Comment updateComment(long id, long profileId, Comment menteeComment)
+            throws ResourceNotFoundException, UnauthorizedException {
+        Optional<Comment> optionalComment = commentRepository.findById(id);
+        if (!optionalComment.isPresent()) {
+            String msg = "Error, Comment with id: " + id + " cannot be deleted. " +
```
Comment (Piumal1999) [inline src/main/java/org/sefglobal/scholarx/service/CommentService.java:93]:
```suggestion
            String msg = "Error, Comment with id: " + id + " cannot be updated. " +
```

**Review by Piumal1999 — APPROVED** [review]
Good job! @anjula-sack 
Thanks for getting this done

**Pull request merged**